rdkPlugins/Logging/source/JournaldSink.cpp

```diff
@@ -0,0 +1,179 @@
+        if (!priority.empty())
+        {
+            const std::map<std::string, int> options =
+                {
+                    {"LOG_EMERG", 0},
+                    {"LOG_ALERT", 1},
+                    {"LOG_CRIT", 2},
+                    {"LOG_ERR", 3},
+                    {"LOG_WARNING", 4},
+                    {"LOG_NOTICE", 5},
+                    {"LOG_INFO", 6},
+                    {"LOG_DEBUG", 7}};
+
+            auto it = options.find(priority);
+            if (it != options.end())
+            {
+                logPriority = it->second;
+            }
+            else
+            {
+                AI_LOG_WARN("Could not parse journald priority - using LOG_INFO");
+            }
+        }
+    }
+
+    mJournaldSteamFd = sd_journal_stream_fd(mContainerId.c_str(), logPriority, 1);
+    if (mJournaldSteamFd < 0)
+    {
+        AI_LOG_SYS_ERROR(-mJournaldSteamFd, "Failed to create journald stream fd");
+
+        // Just use /dev/null instead
+        mJournaldSteamFd = open("/dev/null", O_CLOEXEC | O_WRONLY);
+    }
+
+    AI_LOG_FN_EXIT();
+}
+
+JournaldSink::~JournaldSink()
+{
+    if (mJournaldSteamFd > 0)
+    {
+        if (close(mJournaldSteamFd) < 0)
+        {
+            AI_LOG_SYS_ERROR(errno, "Failed to close journald stream");
+        }
+    }
+}
+
+void JournaldSink::SetLogOptions(const IDobbyRdkLoggingPlugin::LoggingOptions& options)
+{
+    std::lock_guard<std::mutex> locker(mLock);
+    mLoggingOptions = options;
+}
+
+void JournaldSink::DumpLog(const int bufferFd)
+{
+    std::lock_guard<std::mutex> locker(mLock);
+
+    ssize_t ret;
+    memset(mBuf, 0, sizeof(mBuf));
+
+    while (true)
+    {
+        ret = read(bufferFd, mBuf, sizeof(mBuf));
+        if (ret <= 0)
+        {
+            break;
+        }
+
+        write(mJournaldSteamFd, mBuf, ret);
+    }
+}
+
+void JournaldSink::process(const std::shared_ptr<AICommon::IPollLoop> &pollLoop, uint32_t events)
+{
+    std::lock_guard<std::mutex> locker(mLock);
+
+    // Got some data, yay
+    if (events & EPOLLIN)
+    {
+        ssize_t ret;
+        memset(mBuf, 0, sizeof(mBuf));
+
+        while (true)
+        {
+            ret = TEMP_FAILURE_RETRY(read(mLoggingOptions.pttyFd, mBuf, sizeof(mBuf)));
+            if (ret < 0)
+            {
+                // We've reached the end of the data we can read so we're done here
+                if (errno == EWOULDBLOCK)
+                {
+                    return;
+                }
+
+                // Something went wrong whilst reading
+                AI_LOG_SYS_ERROR(errno, "Read from container %s tty failed", mContainerId.c_str());
+                return;
+            }
+
+            if (write(mJournaldSteamFd, mBuf, ret) < 0)
+            {
+                AI_LOG_SYS_ERROR(errno, "Write to journald stream failed");
+            }
+        }
+
+        return;
+    }
+
+    // Container shutdown
+    if (events & EPOLLHUP)
```

Comment:
There is no way to get to this point if previous is true, I would add use "else if" for clarity.